test/common/testlib.py

```diff
@@ -654,6 +662,94 @@ def snapshot(self, title, label=None):
 
+    def assert_pixels(self, selector, key, ignore=[]):
+        """Compare the given element with its reference"""
+
+        if Image and self.pixels_label and self.cdp and self.cdp.valid:
+            rect = self.call_js_func('ph_element_clip', selector)
+
+            @dataclass
+            class Rect:
+                x0: int
+                y0: int
+                x1: int
+                y1: int
+
+            def relative_clip(sel):
+                try:
+                    r = self.call_js_func('ph_element_clip', sel)
+                    return Rect(r['x'] - rect['x'],
+                                r['y'] - rect['y'],
+                                r['x'] - rect['x'] + r['width'],
+                                r['y'] - rect['y'] + r['height'])
+                except Error:
+                    return None
+
+            ignore_rects = list(map(relative_clip, ignore))
+            base = self.pixels_label + "-" + key
+            filename = base + "-pixels.png"
+            ref_filename = os.path.join(TEST_DIR, "reference", filename)
+            ret = self.cdp.invoke("Page.captureScreenshot", clip=rect, no_trace=True)
+            if "data" in ret:
+                png_now = base64.standard_b64decode(ret["data"])
+                png_ref = os.path.exists(ref_filename) and open(ref_filename, "rb").read()
+                if not png_ref:
+                    with open(filename, 'wb') as f:
+                        f.write(png_now)
+                    attach(filename)
+                    print("New pixel test reference " + filename)
+                    self.failed_pixel_tests += 1
+                else:
+                    img_now = Image.open(io.BytesIO(png_now)).convert("RGBA")
+                    img_ref = Image.open(io.BytesIO(png_ref)).convert("RGBA")
+                    img_delta = img_ref.copy()
+
+                    def ignoreable_coord(x, y):
+                        for r in ignore_rects:
+                            if r and x >= r.x0 and x < r.x1 and y >= r.y0 and y < r.y1:
+                                return True
+                        return False
+
+                    def ignoreable_change(a, b):
+                        return abs(a[0] - b[0]) < 2 and abs(a[1] - b[1]) < 2 and abs(a[1] - b[1]) < 2
+
+                    def img_eq(ref, now, delta):
+                        # This is slow but exactly what we want.
+                        # ImageMath might be able to speed this up.
+                        if ref.size != now.size:
+                            return False
+                        data_ref = ref.load()
+                        data_now = now.load()
+                        data_delta = delta.load()
+                        result = True
+                        width, height = ref.size
+                        for y in range(height):
+                            for x in range(width):
+                                if data_ref[x, y][3] == 255 and data_ref[x, y] != data_now[x, y]:
```

Comment:
So this ignores everything with just a wee bit of alpha, right? Can you please add a comment about the high-level comparison strategy, so that we will still know a year from now?

Reply:
Done.